### PR TITLE
gh-100288: Specialise LOAD_ATTR_METHOD for managed dictionaries

### PR DESCRIPTION
**Pull request by Fidget-Spinner**
This specialisation targets the following code:
```
class A:
 def b(s): ...

a = A()
a.__dict__
def f():
 for _ in range(10):
  a.b() # This LOAD_ATTR_METHOD was previously not specialised.

f()
```

* Issue: gh-100288

### CODE REVIEW

**Comment by Fidget-Spinner** [comment]
Interesting, I can create a reproducer:
```
import dis
import configparser as c

p = c.ConfigParser()
def f():
 for _ in range(10):
  print(p.optionxform("Test"))

f()
p.optionxform = lambda a:a
f() # supposed to be "Test" but prints "test"
```

File: Python/bytecodes.c [hunk 1]
```diff
@@ -2726,6 +2726,29 @@ dummy_func(
...
+            PyObject *self = TOP();
+            PyTypeObject *self_cls = Py_TYPE(self);
+            _PyLoadMethodCache *cache = (_PyLoadMethodCache *)next_instr;
+            uint32_t type_version = read_u32(cache->type_version);
+            assert(type_version != 0);
+            DEOPT_IF(self_cls->tp_version_tag != type_version, LOAD_ATTR)
+            assert(self_cls->tp_flags & Py_TPFLAGS_MANAGED_DICT);
+            PyDictOrValues dorv = *_PyObject_DictOrValuesPointer(self);
+            DEOPT_IF(_PyDictOrValues_IsValues(dorv), LOAD_ATTR);
+            PyDictKeysObject *keys = ((PyHeapTypeObject *)self_cls)->ht_cached_keys;
```
Comment (markshannon) [inline Python/bytecodes.c]: This is the wrong keys. You need to check the object's dictionary's keys.

**Review by markshannon — COMMENTED** [review]
Is this case common enough to justify the extra instruction?

(Thinking aloud here)
And is this best approach? Can we modify the layout of objects so that it doesn't matter if the `__dict__` has been materialized, provided the keys haven't changed?

File: Python/bytecodes.c (hunk 1 above)
Comment (Fidget-Spinner) [reply Python/bytecodes.c]: Yes I just realised (see my message to you on discord :()

Comment (Fidget-Spinner) [reply Python/bytecodes.c]: Hmm this is what I get for copying the code above.
I've been investigating around a bit, and wouldn't this mean that the code here is also wrong? Or are objects with managed values handled differently?
I started to doubt myself after reading the code for LOAD_ATTR_METHOD_WITH_VALUES https://github.com/python/cpython/blob/9e7862b20bfea6e82af8deffcc7c593e5bb6753f/Python/bytecodes.c#L2705-L2726

**Comment by Fidget-Spinner** [comment]
> Is this case common enough to justify the extra instruction?

It's the highest specialisation for LOAD_ATTR we currently have (35.5% of 16.2% specialisation failures).

File: Python/bytecodes.c (hunk 1 above)
Comment (Fidget-Spinner) [reply Python/bytecodes.c]: Ah wait, I forgot that managed values means the object has no instance dict 🤦 . My bad.

**Comment by markshannon** [comment]
> It's the highest specialisation failure for LOAD_ATTR we currently have (35.5% of 16.2% specialisation failures).

Can't argue with the data :slightly_smiling_face: 

OK, let's go with this approach. We can always looks for ways to improve the object layout later.

**Comment by Fidget-Spinner** [comment]
@markshannon is it alright if I merge this?

**Comment by markshannon** [comment]
> @markshannon is it alright if I merge this?

Yes, but I'm not pressing any merge buttons just before Christmas :slightly_smiling_face:

**Comment by Fidget-Spinner** [comment]
> Yes, but I'm not pressing any merge buttons just before Christmas 🙂

HAHA right. I'll press it and if anything goes wrong it's 100% on me to fix it :).

**Comment by Fidget-Spinner** [comment]
Well looks like I'm reverting PRs for Christmas... this PR introduced a race condition in `test___xxxsubinterpreters` on Windows. I had to run this with `-F` and it only crashed after 10 runs.
```
Thread 0x00000430 (most recent call first):
  File "D:\a\1\s\Lib\test\test__xxsubinterpreters.py", line 812 in test_create_thread
  File "D:\a\1\s\Lib\unittest\case.py", line 579 in _callTestMethod
  File "D:\a\1\s\Lib\unittest\case.py", line 623 in run
  File "D:\a\1\s\Lib\unittest\case.py", line 678 in __call__
  File "D:\a\1\s\Lib\unittest\suite.py", line 122 in run
  File "D:\a\1\s\Lib\unittest\suite.py", line 84 in __call__
  File "D:\a\1\s\Lib\unittest\suite.py", line 122 in run
  File "D:\a\1\s\Lib\unittest\suite.py", line 84 in __call__
  File "D:\a\1\s\Lib\unittest\suite.py", line 122 in run
  File "D:\a\1\s\Lib\unittest\suite.py", line 84 in __call__
  File "D:\a\1\s\Lib\unittest\runner.py", line 208 in run
  File "D:\a\1\s\Lib\test\support\__init__.py", line 1100 in _run_suite
  File "D:\a\1\s\Lib\test\support\__init__.py", line 1226 in run_unittest
  File "D:\a\1\s\Lib\test\libregrtest\runtest.py", line 281 in _test_module
  File "D:\a\1\s\Lib\test\libregrtest\runtest.py", line 317 in _runtest_inner2
  File "D:\a\1\s\Lib\test\libregrtest\runtest.py", line 360 in _runtest_inner
  File "D:\a\1\s\Lib\test\libregrtest\runtest.py", line 235 in _runtest
  File "D:\a\1\s\Lib\test\libregrtest\runtest.py", line 265 in runtest
  File "D:\a\1\s\Lib\test\libregrtest\main.py", line 353 in rerun_failed_tests
  File "D:\a\1\s\Lib\test\libregrtest\main.py", line 756 in _main
  File "D:\a\1\s\Lib\test\libregrtest\main.py", line 711 in main
  File "D:\a\1\s\Lib\test\libregrtest\main.py", line 775 in main
  File "D:\a\1\s\Lib\test\__main__.py", line 2 in <module>
  File "<frozen runpy>", line 88 in _run_code
  File "<frozen runpy>", line 198 in _run_module_as_main
##[error]Cmd.exe exited with code '-1073741819'.

```

**Pull request merged**